File: MINCO/src/decision_math/progressive_hedging.py
"""Small-instance Progressive Hedging reference implementation.

This is an independently executable algorithmic check for MINCO's primary
Julia/JuMP Progressive Hedging implementation. It supports the one-hospital,
one-period benchmark where first-stage decisions are surge activation and flex
staff blocks.
"""
from __future__ import annotations

from dataclasses import dataclass
from itertools import product

import numpy as np

from .stochastic_milp_oracle import StochasticCapacityInstance


@dataclass(frozen=True)
class ProgressiveHedgingResult:
    consensus_surge: int
    consensus_flex: int
    iterations: int
    converged: bool
    scenario_decisions: np.ndarray
    residual_history: tuple[float, ...]


def _scenario_recourse_loss(inst: StochasticCapacityInstance, wi: int, surge: int, flex: int) -> float:
    if len(inst.hospitals) != 1 or len(inst.periods) != 1:
        raise ValueError("Reference PH supports one hospital and one period")
    physical = inst.base_beds[0] + surge * inst.surge_beds[0] + flex * inst.flex_beds_per_block[0]
    safe = inst.safe_fraction[0] * physical
    demand = float(inst.demand[wi, 0, 0])
    max_def = int(round(float(inst.elective_deferral_limit[0, 0])))
    best = float("inf")
    for defer in range(max_def + 1):
        treated = max(0.0, demand - defer)
        unsafe = max(0.0, treated - safe)
        boarding = max(0.0, treated - physical)
        loss = inst.defer_cost * defer + inst.unsafe_cost * unsafe + inst.boarding_cost * boarding
        best = min(best, float(loss))
    return best


def progressive_hedging_single_hospital(
    inst: StochasticCapacityInstance,
    *,
    rho: float = 12.0,
    max_iter: int = 100,
    tolerance: float = 1e-9,
) -> ProgressiveHedgingResult:
    inst.validate()
    if rho <= 0 or max_iter <= 0:
        raise ValueError("rho and max_iter must be positive")
    w_count = len(inst.scenario_probabilities)
    candidates = np.array(
        list(product([0, 1], range(int(inst.max_flex_blocks[0]) + 1))),
        dtype=float,
    )
    first_cost = candidates[:, 0] * inst.surge_cost[0] + candidates[:, 1] * inst.flex_cost[0]

    multipliers = np.zeros((w_count, 2), dtype=float)
    decisions = np.zeros((w_count, 2), dtype=float)
    # Initial independent scenario solves.
    for wi in range(w_count):
        objectives = np.array([
            first_cost[k] + _scenario_recourse_loss(inst, wi, int(x[0]), int(x[1]))
            for k, x in enumerate(candidates)
        ])
        decisions[wi] = candidates[int(np.argmin(objectives))]

    residuals: list[float] = []
    converged = False
    for iteration in range(1, max_iter + 1):
        xbar = np.average(decisions, axis=0, weights=inst.scenario_probabilities)
        new_decisions = np.empty_like(decisions)
        for wi in range(w_count):
            values = []
            for k, candidate in enumerate(candidates):
                augmented = (
                    first_cost[k]
                    + _scenario_recourse_loss(inst, wi, int(candidate[0]), int(candidate[1]))
                    + float(multipliers[wi] @ candidate)
                    + 0.5 * rho * float(np.sum((candidate - xbar) ** 2))
                )
                values.append(augmented)
            new_decisions[wi] = candidates[int(np.argmin(values))]
        decisions = new_decisions
        xbar = np.average(decisions, axis=0, weights=inst.scenario_probabilities)
        residual = float(np.sqrt(np.sum(inst.scenario_probabilities[:, None] * (decisions - xbar) ** 2)))
        residuals.append(residual)
        if residual <= tolerance:
            converged = True
            break
        multipliers += rho * (decisions - xbar)

    # PH consensus is integral only when all scenario decisions agree.
    consensus = np.rint(np.average(decisions, axis=0, weights=inst.scenario_probabilities)).astype(int)
    return ProgressiveHedgingResult(
        consensus_surge=int(consensus[0]),
        consensus_flex=int(consensus[1]),
        iterations=iteration,
        converged=converged,
        scenario_decisions=decisions.astype(int),
        residual_history=tuple(residuals),
    )

File: MINCO/julia/src/progressive_hedging.jl
"""Progressive Hedging for scenario-decomposed MINCO first-stage decisions.

The implementation solves one scenario MILP/MIQP per iteration and penalizes
non-anticipativity deviations in surge and flex decisions.
"""
function _solve_ph_scenario(data::CapacityData, w::Int, xbar_surge, xbar_flex, u_surge, u_flex, rho::Float64; optimizer)
    H, T = length(data.hospitals), length(data.periods)
    model = Model(optimizer)
    set_silent(model)
    @variable(model, surge[1:H,1:T], Bin)
    @variable(model, flex[1:H,1:T] >= 0, Int)
    @variable(model, defer[1:H,1:T] >= 0, Int)
    @variable(model, transfer[1:H,1:H,1:T] >= 0, Int)
    @variable(model, unsafe[1:H,1:T] >= 0)
    @variable(model, boarding[1:H,1:T] >= 0)
    @constraint(model, [h=1:H,t=1:T], flex[h,t] <= data.max_flex_blocks[h])
    @constraint(model, [h=1:H,t=1:T], defer[h,t] <= data.deferral_limit[h,t])
    @constraint(model, [i=1:H,j=1:H,t=1:T; i != j], transfer[i,j,t] <= data.transfer_capacity[i,j,t])
    @constraint(model, [h=1:H,t=1:T], transfer[h,h,t] == 0)
    @expression(model, capacity[h=1:H,t=1:T], data.base_beds[h] + data.surge_beds[h]*surge[h,t] + data.flex_beds_per_block[h]*flex[h,t])
    @expression(model, treated[h=1:H,t=1:T], data.demand[w,h,t] - defer[h,t] + sum(transfer[j,h,t] for j=1:H if j!=h) - sum(transfer[h,j,t] for j=1:H if j!=h))
    @constraint(model, [h=1:H,t=1:T], boarding[h,t] >= treated[h,t] - capacity[h,t])
    @constraint(model, [h=1:H,t=1:T], unsafe[h,t] >= treated[h,t] - data.safe_fraction[h]*capacity[h,t])
    @expression(model, base_cost,
        sum(data.surge_cost[h]*surge[h,t] + data.flex_cost[h]*flex[h,t] for h=1:H,t=1:T)
        + data.defer_cost*sum(defer)
        + data.transfer_cost*sum(transfer[i,j,t] for i=1:H,j=1:H,t=1:T if i!=j)
        + data.unsafe_cost*sum(unsafe)
        + data.boarding_cost*sum(boarding))
    @expression(model, penalty,
        sum(u_surge[h,t]*surge[h,t] + 0.5*rho*(surge[h,t]-xbar_surge[h,t])^2 for h=1:H,t=1:T)
        + sum(u_flex[h,t]*flex[h,t] + 0.5*rho*(flex[h,t]-xbar_flex[h,t])^2 for h=1:H,t=1:T))
    @objective(model, Min, base_cost + penalty)
    optimize!(model)
    has_values(model) || error("Progressive Hedging scenario $w failed: $(termination_status(model))")
    return value.(surge), value.(flex), objective_value(model)
end

function progressive_hedging(data::CapacityData; optimizer, rho::Float64=20.0, tolerance::Float64=1e-3, max_iterations::Int=60)
    _validate(data)
    H,T,W = length(data.hospitals), length(data.periods), length(data.probabilities)
    x_surge = zeros(W,H,T); x_flex = zeros(W,H,T)
    u_surge = zeros(W,H,T); u_flex = zeros(W,H,T)
    xbar_surge = zeros(H,T); xbar_flex = zeros(H,T)
    residual_history = Float64[]
    objective_history = Float64[]

    # Standard PH initialization: solve every scenario independently before
    # introducing the augmented non-anticipativity penalty.
    for w in 1:W
        x_surge[w,:,:], x_flex[w,:,:], _ = _solve_ph_scenario(
            data, w, xbar_surge, xbar_flex, u_surge[w,:,:], u_flex[w,:,:], 0.0; optimizer=optimizer)
    end
    xbar_surge .= 0.0
    xbar_flex .= 0.0
    for w in 1:W
        xbar_surge .+= data.probabilities[w] .* x_surge[w,:,:]
        xbar_flex .+= data.probabilities[w] .* x_flex[w,:,:]
    end

    for iteration in 1:max_iterations
        objectives = zeros(W)
        for w in 1:W
            x_surge[w,:,:], x_flex[w,:,:], objectives[w] = _solve_ph_scenario(
                data, w, xbar_surge, xbar_flex, u_surge[w,:,:], u_flex[w,:,:], rho; optimizer=optimizer)
        end
        xbar_surge .= 0.0
        xbar_flex .= 0.0
        for w in 1:W
            xbar_surge .+= data.probabilities[w] .* x_surge[w,:,:]
            xbar_flex .+= data.probabilities[w] .* x_flex[w,:,:]
        end
        residual = sqrt(sum(data.probabilities[w]*(sum((x_surge[w,:,:].-xbar_surge).^2)+sum((x_flex[w,:,:].-xbar_flex).^2)) for w=1:W))
        push!(residual_history, residual); push!(objective_history, sum(data.probabilities .* objectives))
        residual <= tolerance && return Dict("converged"=>true,"iterations"=>iteration,"residual_history"=>residual_history,"surge_consensus"=>round.(Int,xbar_surge),"flex_consensus"=>round.(Int,xbar_flex),"scenario_objective_history"=>objective_history)
        for w in 1:W
            u_surge[w,:,:] .+= rho .* (x_surge[w,:,:] .- xbar_surge)
            u_flex[w,:,:] .+= rho .* (x_flex[w,:,:] .- xbar_flex)
        end
    end
    return Dict("converged"=>false,"iterations"=>max_iterations,"residual_history"=>residual_history,"surge_consensus"=>round.(Int,xbar_surge),"flex_consensus"=>round.(Int,xbar_flex),"scenario_objective_history"=>objective_history)
end
